### PR TITLE
feat: support custom blocks to rollback 

### PR DESCRIPTION
**Pull request by cosinlink**
### Description

support custom blocks to rollback 

### Changes

Notable changes:
* support custom blocks to rollback 
* ...

### CODE REVIEW

**Comment by forcodedancing** [comment]
there are some lint issues

File: state/rollback.go [hunk 1]
```diff
@@ -12,7 +12,7 @@ import (
 // Rollback overwrites the current CometBFT state (height n) with the most
 // recent previous state (height n - 1).
 // Note that this function does not affect application state.
-func Rollback(bs BlockStore, ss Store, removeBlock bool) (int64, []byte, error) {
+func Rollback(bs BlockStore, ss Store, removeBlock bool, rollbackBlocks int64) (int64, []byte, error) {
 	invalidState, err := ss.Load()
```
Comment (forcodedancing) [inline state/rollback.go:16]: `ss.Load()` will load the latest state, which is not the state we are expecting.

Comment (forcodedancing) [reply state/rollback.go:16]: We need heavy changes for this function.

File: cmd/cometbft/commands/rollback.go [hunk 2]
```diff
@@ -15,25 +15,27 @@ import (
 )
 
 var removeBlock = false
+var rollbackBlocks = uint(1)
```
Comment (forcodedancing) [inline cmd/cometbft/commands/rollback.go:18]: We also need to limit the max rollback blocks in one command.
If a user wants to rollback many blocks, he/she can run the command several times.

Comment (cosinlink) [reply cmd/cometbft/commands/rollback.go:22]: fixed

File: state/rollback.go (hunk 1 above)
Comment (forcodedancing) [reply state/rollback.go:16]:
```
		NextValidators:              invalidState.Validators,
		Validators:                  invalidState.LastValidators,
```
This should be updated.

Comment (cosinlink) [reply state/rollback.go:15]: fixed to validators loaded from store

**Comment by forcodedancing** [comment]
We need to test validator set change, maybe on dev or qa env.

**Review by forcodedancing — APPROVED** [review]

File: state/store.go [hunk 3]
```diff
@@ -173,7 +177,10 @@ func (store dbStore) Save(state State) error {
 	return store.save(state, stateKey)
 }
 
+// [243 5vals] ==state243(4 vals)=> [244 4vals] => [245 4vals]
```
Comment (forcodedancing) [inline state/store.go:180]: Please clean the codes

File: state/store.go [hunk 4]
```diff
@@ -633,6 +641,41 @@ func (store dbStore) loadConsensusParamsInfo(height int64) (*cmtstate.ConsensusP
...
+	var (
+		emptypb = cmtproto.ConsensusParams{}
+	)
+	paramsInfo, err := store.loadConsensusParamsInfo(height)
+	if err != nil {
+		return 0, fmt.Errorf("could not find consensus params for height #%d: %w", height, err)
+	}
+
+	if paramsInfo.ConsensusParams.Equal(&emptypb) {
+		paramsInfo2, err := store.loadConsensusParamsInfo(paramsInfo.LastHeightChanged)
```
Comment (forcodedancing) [inline state/store.go:654]: If we already get the height, why not just return it?

Comment (cosinlink) [reply state/store.go]: removed the unused code

File: state/store.go (hunk 3 above)
Comment (cosinlink) [reply state/store.go]: cleaned

**Review by zjubfd — APPROVED** [review]

**Review by forcodedancing — APPROVED** [review]

**Pull request merged**